Case Deduplication

proteusvacuum opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
This adds a new case duplication feature which works offline.

Case duplicates are marked based on a rule which are similar to "automatic update rule". A new pillow processor for the xforms pillow picks up changes and marks any updated cases as duplicates based on these rules on submission. Cases marked as duplicates can also be automatically updated.
A report, based on the case list explorer, can be used to download, filter and explore any cases marked as a duplicate.

https://docs.google.com/document/d/1tGXWBub0CymOg06lwreFWj1Tq5IpRbSMzyblcVHCBKY/edit


## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

https://docs.google.com/document/d/1tGXWBub0CymOg06lwreFWj1Tq5IpRbSMzyblcVHCBKY/edit

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
Case Search and the new CASE_DEDUPE flag

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
This will go through QA

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
This adds a new pillow processor to the xform-pillow, but otherwise doesn't touch any existing code. 
The "safety" issue I foresee is a scale problem. I've tested this extensively on staging with a domain with 500,000 cases and it has performed "fine". There is an "off switch" where we can disable the DEDUPLICATION type rules immediately if there are any issues, and the code will short circuit. 

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

proteusvacuum committed:
Add case dedup toggle

proteusvacuum committed:
Add criteria match for deduplication workflow

proteusvacuum committed:
Add casededuplication action models

proteusvacuum committed:
Find duplicates for case

proteusvacuum committed:
Add CaseDeduplicationAction

proteusvacuum committed:
Add more test stubs

proteusvacuum committed:
Add TODO: Batch things, friend

proteusvacuum committed:
Don't allow closing cases

proteusvacuum committed:
Fetch and update cases in bulk

proteusvacuum committed:
Save all duplicates in bulk

proteusvacuum committed:
Delete CaseDuplicates

proteusvacuum committed:
Case Dedupe Pillow

proteusvacuum committed:
WIP backfill

proteusvacuum committed:
Add CaseDedupe models to Django Admin

proteusvacuum committed:
Add tests

proteusvacuum committed:
Use BigInt and add uniqueness constraint

proteusvacuum committed:
Update tests to be more rigourous

proteusvacuum committed:
Optimize creation and deletion of duplicates

proteusvacuum committed:
Use toggle in pillow

proteusvacuum committed:
Short circuit on form XMLNS for case updates

proteusvacuum committed:
Clarify a method

proteusvacuum committed:
Add domain deletion models

proteusvacuum committed:
Don't update unique cases

kaapstorm commented on corehq/apps/data_interfaces/admin.py:
```diff
@@ -47,5 +51,49 @@ class CaseRuleSubmissionAdmin(admin.ModelAdmin):
+
+    def dedupe_action_link(self, obj):
+        url = reverse("admin:data_interfaces_casededuplicationactiondefinition_change", args=[obj.definition.id])
+        return format_html(f'<a href={url}>{str(obj.definition)}</a>')
```
What does `str()` do here?

kaapstorm reviewed (approved):
This is looking great.

I like the test docstrings. I think they explain what the test will verify better than a method name.

proteusvacuum commented on corehq/apps/data_interfaces/deduplication.py:
```diff
@@ -0,0 +1,22 @@
+    for case_property_name in case_properties:
+        es = es.case_property_query(
+            case_property_name,
+            case.get_case_property(case_property_name) or '',
```
I believe it will always be a string

snopoke commented on corehq/apps/data_interfaces/tests/test_case_deduplication.py:
```diff
@@ -0,0 +1,486 @@
+class CaseRuleCriteriaTest(TestCase):
+
+    @classmethod
+    def setUpClass(cls, *args, **kwargs):
```
since this is an entirely a SQL test you can use `setUpTestData` (which doesn't require calling super) and leave out the teardown (SQL transactions will take care of cleanup)

proteusvacuum replied:
Cool! I didn't know about that. Have updated.

snopoke commented on corehq/apps/data_interfaces/tests/test_case_deduplication.py:
```diff
@@ -0,0 +1,486 @@
+        ])
+
+
+@override_settings(TESTS_SHOULD_USE_SQL_BACKEND=True)
```
this shouldn't be necessary anymore https://github.com/dimagi/commcare-hq/pull/30411

proteusvacuum replied:
👍🏽

snopoke reviewed (commented):
few small comments but mostly this looks good

proteusvacuum commented on corehq/apps/data_interfaces/deduplication.py:
```diff
@@ -18,5 +20,4 @@ def find_duplicate_ids_for_case(domain, case, case_properties, include_closed=Fa
         )
-
-    return es.get_ids()
+    return [CommCareCase.wrap(flatten_result(hit)) for hit in es.run().hits]
```
This is a good question, and I think was due to laziness on my part when I first wrote this when creating the case search index a while ago. It's a bit of a lift to convert this to use `CommCareCaseSQL`, and I'm not even sure that is the right model to use here. For now, I'm going to leave this as is, and will address this separately as I think it deserves its own PR.

millerdev replied:
FYI, I will be attempting to remove `CommCareCase` sometime soon. It would be nice if we were not writing new code that uses it.

proteusvacuum committed:
use set methods

proteusvacuum committed:
CaseDuplicate transactions should be atomic

proteusvacuum committed:
Remove TESTS_SHOULD_USE_SQL_BACKEND and use setUpTestData

proteusvacuum committed:
Remove TODO

proteusvacuum commented on corehq/apps/data_interfaces/models.py:
```diff
@@ -874,6 +876,28 @@ def when_case_matches(self, case, rule):
+    ALL = "ALL"
+    CHOICES = (
+        (ANY, ANY),
+        (ALL, ALL),
```
No, these will show up differently in the UI

proteusvacuum commented on corehq/apps/data_interfaces/tests/test_case_deduplication.py:
```diff
@@ -196,13 +195,6 @@ def test_integration_test(self):
         """Don't mock the find_duplicate_ids response to make sure it works
         """
 
-    def test_close_case_closes_cases(self):
```
Yeah, I'm going to add similar tests when I build out the UI.

proteusvacuum commented on corehq/apps/data_interfaces/models.py:
```diff
@@ -866,6 +897,184 @@ def when_case_matches(self, case, rule):
+
+        any_match = (
+            self.match_type == CaseDeduplicationMatchTypeChoices.ANY
+            and len(case_properties - definition_properties) < len(case_properties)
```
Good point. Updated.

proteusvacuum commented on corehq/apps/data_interfaces/models.py:
```diff
@@ -866,6 +897,184 @@ def when_case_matches(self, case, rule):
+        )
+
+    @classmethod
+    def bulk_create_duplicate_relationships(cls, action, initial_case, duplicate_cases):
```
Yes. Updated.

proteusvacuum committed:
Deduplication Rule List View

proteusvacuum committed:
Create View

proteusvacuum committed:
Edit View

proteusvacuum committed:
Refactor: initiate_messaging_rule_run -> initiate_rule_run

proteusvacuum committed:
Update mgmt command to accept deduplicate rules

proteusvacuum committed:
Backfill deduplicates after creating

proteusvacuum committed:
List deduplication rules template

TODO: DRY this

proteusvacuum committed:
Atomic updates

proteusvacuum committed:
Remove trailing commas

proteusvacuum committed:
Fetch cases from the DB when updating so they only get updated once

proteusvacuum committed:
Update test with find_duplicate_case_ids

proteusvacuum committed:
jslint

proteusvacuum committed:
Remove last_run check

For the backfill process, this variable is meaningless since we always want to
run on all cases. For individual cases, the rule only ever gets called on
creates and updates so this should be an extraneous check. Updating `last_run`
on every form submission would also be confusing.

proteusvacuum committed:
Backfill duplicates in a separate task

proteusvacuum committed:
Use case_rule_queue instead

proteusvacuum committed:
Error handling in task

proteusvacuum committed:
Remove unused imports

proteusvacuum committed:
Small UI tweak to align items correctly

proteusvacuum committed:
Add deduplicate cases report

proteusvacuum committed:
Require at least one matching property

Charl1996 reviewed (approved):
This looks like a fun feature to have developed!

snopoke reviewed (commented):
This looks great. I've made a few comments and I was also wondering if we need to handle case deletions (ie. that result from deleting a user).

proteusvacuum committed:
Revert "Refactor: initiate_messaging_rule_run -> initiate_rule_run"

This reverts commit 5f6d541c091ad88cedb51a4fd9d22a3ce162ff26.

I didn't end up using this method so revert the refactor

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/dedupe

proteusvacuum committed:
Get the action from the rule

proteusvacuum committed:
Split reset and backfill rules and move code to utils

proteusvacuum committed:
Re-backfill when activating a rule

proteusvacuum committed:
Delete CaseDuplicates when rule is deleted

proteusvacuum committed:
Remove last run column

proteusvacuum committed:
Always have default columns first in Duplicates Explorer

proteusvacuum committed:
Lock rule for editing before backfilling is queued

proteusvacuum committed:
Show duplicate count and status on list page

proteusvacuum committed:
Remove admin models

snopoke committed:
Merge branch 'master' into fr/dedupe

snopoke committed:
fix merge with master class name change

proteusvacuum committed:
Allow exploring inactive rules

proteusvacuum committed:
Rewording

proteusvacuum committed:
Show errors when properties are used multiple times

proteusvacuum committed:
Reload the page

proteusvacuum committed:
Form topic no longer exists

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/dedupe

proteusvacuum committed:
Prevent updating reserved case properties

proteusvacuum committed:
Don't explicitly delete cases in tests

proteusvacuum committed:
Try/finally when running dedupe rule

proteusvacuum committed:
Use ValueError

proteusvacuum committed:
Use an f-string

proteusvacuum committed:
Remove_duplicates_for_action method

proteusvacuum committed:
Only update cases if there are updates to carry out

proteusvacuum committed:
Remove unneeded check

proteusvacuum committed:
Append rule name to xmlns

proteusvacuum commented on corehq/apps/data_interfaces/admin.py:
```diff
@@ -47,5 +51,49 @@ class CaseRuleSubmissionAdmin(admin.ModelAdmin):
+class CaseDuplicateAdmin(admin.ModelAdmin):
+    model = CaseDuplicate
+    search_fields = ['case_id']
+    list_filter = ['action', 'case_id']
```
I've removed this UI completely now

proteusvacuum commented on corehq/apps/data_interfaces/deduplication.py:
```diff
@@ -0,0 +1,50 @@
+        return
+
+    if rule.workflow != AutomaticUpdateRule.WORKFLOW_DEDUPLICATE:
+        raise AttributeError
```
https://github.com/dimagi/commcare-hq/pull/30425/commits/2fd1690ad0ff195659326567ea30451a2b4fd1f8

proteusvacuum commented on corehq/apps/data_interfaces/models.py:
```diff
@@ -766,13 +775,13 @@ def set_properties_to_update(self, properties):
         for p in properties:
             if not isinstance(p, self.PropertyDefinition):
-                raise ValueError("Expected UpdateCaseDefinition.PropertyDefinition")
+                raise ValueError("Expected {}.PropertyDefinition".format(self.__class__.__name__))
```
https://github.com/dimagi/commcare-hq/pull/30425/commits/0a25bdddbd6b137e463ea8c5efcd5ea20cacebe7

proteusvacuum commented on corehq/apps/data_interfaces/models.py:
```diff
@@ -866,6 +892,194 @@ def when_case_matches(self, case, rule):
+    def _update_cases(self, domain, rule, duplicate_case_ids):
+        """Updates all the duplicate cases according to the rule
+        """
+        duplicate_cases = CaseAccessors(domain).get_cases(list(duplicate_case_ids))
```
Great point: https://github.com/dimagi/commcare-hq/pull/30425/commits/b3644099a691d6a7a9b031f65c58986de21d1549

proteusvacuum commented on corehq/apps/data_interfaces/pillow.py:
```diff
@@ -0,0 +1,64 @@
+        if action.definition.properties_fit_definition(changed_properties):
+            try:
+                case = CaseAccessors(domain).get_case(case_id)
+            except CaseNotFound:
```
I guess never? I don't remember why I added this, but I've removed it now: https://github.com/dimagi/commcare-hq/pull/30425/commits/f6870f7e3ace4d46b70d7149f06ccdd7c1900421

proteusvacuum commented on corehq/apps/data_interfaces/tasks.py:
```diff
@@ -57,6 +61,58 @@ def _progress_tracker(current, total):
+    AutomaticUpdateRule.objects.filter(pk=rule.pk).update(
+        locked_for_editing=False,
+        last_run=now,
+    )
```
https://github.com/dimagi/commcare-hq/pull/30425/commits/4a48a99d47ba71e7bacc961cb8d916a617810655

proteusvacuum commented on corehq/apps/data_interfaces/tests/test_case_deduplication.py:
```diff
@@ -0,0 +1,478 @@
+        self.factory = CaseFactory(self.domain)
+
+    def tearDown(self):
+        FormProcessorTestUtils.delete_all_cases()
```
https://github.com/dimagi/commcare-hq/pull/30425/commits/0ef456d0339d92dc82cd5c700fb98362370c2dd7

proteusvacuum commented on corehq/apps/data_interfaces/tests/test_case_deduplication.py:
```diff
@@ -0,0 +1,478 @@
+
+    @patch("corehq.apps.data_interfaces.models.find_duplicate_case_ids")
+    def test_pillow_processes_changes(self, find_duplicate_cases_mock):
+        kafka_sec = get_multi_topic_offset([topics.FORM, topics.FORM_SQL])
```
https://github.com/dimagi/commcare-hq/pull/30425/commits/f758832351c3678d4a702f4e26ad8f32bdbcaba0

proteusvacuum commented on corehq/apps/data_interfaces/models.py:
```diff
@@ -68,6 +66,7 @@
 
 ALLOWED_DATE_REGEX = re.compile(r'^\d{4}-\d{2}-\d{2}')
 AUTO_UPDATE_XMLNS = 'http://commcarehq.org/hq_case_update_rule'
+DEDUPE_XMLNS = 'http://commcarehq.org/hq_case_deduplication_rule'
```
https://github.com/dimagi/commcare-hq/pull/30425/commits/0bc4c7df40e901fea643fbfb37045a4e6f176705

snopoke replied:
I think this wasn't a great suggestion and we should have gone with `form.meta.deviceID` to store the specific rule which is the recommended approach: https://github.com/dimagi/commcare-hq/blob/5cdc537484fe6db18f533bafef70256caa861281/corehq/apps/hqcase/utils.py#L42-L52

proteusvacuum replied:
@snopoke Do you think it's worth changing it at this point?

proteusvacuum commented on corehq/apps/data_interfaces/models.py:
```diff
@@ -866,6 +892,194 @@ def when_case_matches(self, case, rule):
+        )
+
+    @classmethod
+    def bulk_create_duplicate_relationships(cls, action, initial_case, duplicate_case_ids):
```
This is in an atomic block when it is called, https://github.com/dimagi/commcare-hq/blob/fr%2Fdedupe/corehq/apps/data_interfaces/models.py#L973-L976 
Do you think that suffices?

proteusvacuum committed:
Delete CaseDuplicates when cases are deleted

proteusvacuum commented:
@snopoke thanks for the thorough review! I believe I've now responded to all the comments. This has also now passed QA.

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/dedupe

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/dedupe

snopoke commented on corehq/apps/data_interfaces/deduplication.py:
```diff
@@ -106,3 +107,11 @@ def backfill_deduplicate_rule(domain, rule):
+
+
+def is_dedupe_xmlns(xmlns):
+    return xmlns.split("__")[0] == DEDUPE_XMLNS
```
nit: `xmlns.startswith(DEDUPE_XMLNS)` seems simpler

snopoke commented on corehq/apps/data_interfaces/deduplication.py:
```diff
@@ -106,3 +107,11 @@ def backfill_deduplicate_rule(domain, rule):
+
+
+def get_dedupe_xmlns(rule):
+    return f"{DEDUPE_XMLNS}__{rule.name}-{rule.case_type}"
```
I'm not sure what a name will look like but it would probably be safest to slugify it first

proteusvacuum committed:
Sluglify rule name

proteusvacuum committed:
Add duplicate limit test

proteusvacuum committed:
Don't allow duplicate deduplicate names

snopoke commented:
🚀
